apps/gdalalg_dispatcher.h

```diff
@@ -0,0 +1,242 @@
+    }
+
+    auto poDSFromRaster = m_rasterDispatcher->GetDatasetRef();
+    // cppcheck-suppress knownConditionTrueFalse
+    if (poDSFromRaster)
+    {
+        m_vectorDispatcher->SetDataset(poDSFromRaster);
+    }
+
+    std::vector<std::string> argsWithoutInput;
+    bool skipNext = false;
+    for (const auto &arg : args)
+    {
+        if (arg == "-i" || arg == "--input")
+        {
+            skipNext = true;
+        }
+        else if (!skipNext)
+        {
+            if (!STARTS_WITH(arg.c_str(), "--input=") &&
+                !(poDSFromRaster && arg == poDSFromRaster->GetDescription()))
+            {
+                argsWithoutInput.push_back(arg);
+            }
+        }
+        else
+        {
+            skipNext = false;
+        }
+    }
+
+    {
+        CPLErrorStateBackuper oErrorHandler(CPLQuietErrorHandler);
+        ok = m_vectorDispatcher->ParseCommandLineArguments(argsWithoutInput);
+    }
+    if (ok)
+    {
+        m_selectedSubAlg = m_vectorDispatcher.get();
+        std::vector<std::string> callPath(m_callPath);
+        callPath.push_back("vector");
+        m_selectedSubAlg->SetCallPath(callPath);
+
+        return true;
+    }
+
+    bool ret = false;
+    for (const auto &arg : args)
+    {
+        VSIStatBufL sStat;
+        if (VSIStatL(arg.c_str(), &sStat) == 0)
+        {
+            auto poDS =
+                std::unique_ptr<GDALDataset>(GDALDataset::Open(arg.c_str()));
+            if (poDS)
+            {
+                if (poDS->GetRasterCount() > 0 ||
+                    poDS->GetMetadata("SUBDATASETS"))
+                {
+                    if (poDS->GetLayerCount() != 0)
+                    {
+                        m_showUsage = false;
+                        CPLError(CE_Failure, CPLE_AppDefined,
+                                 "'%s' has both raster and vector content. "
+                                 "Please use 'gdal raster %s' or 'gdal "
+                                 "vector %s'.",
+                                 poDS->GetDescription(), GetName().c_str(),
+                                 GetName().c_str());
+                        return false;
+                    }
+                    m_rasterDispatcher = std::make_unique<RasterDispatcher>();
+                    auto poDSRaw = poDS.get();
+                    m_rasterDispatcher->SetDataset(poDS.release());
+                    poDSRaw->Release();
+                    m_selectedSubAlg = m_rasterDispatcher.get();
+                    std::vector<std::string> callPath(m_callPath);
+                    callPath.push_back("raster");
+                    m_selectedSubAlg->SetCallPath(callPath);
+                    ret = m_selectedSubAlg->ParseCommandLineArguments(
+                        argsWithoutInput);
+                    break;
+                }
+                else if (poDS->GetLayerCount() != 0)
+                {
+                    m_vectorDispatcher = std::make_unique<VectorDispatcher>();
```

Comment:
yes, similar indeed, but it would require a templated helper to make it generic, and I'm not a big fan of templates for just a one-time repetition of a few lines... (because I suck at writing them)